schema-layer/data-structures/generics.md

```diff
@@ -0,0 +1,109 @@
+
+  * The IPLD Data-Model. There are codec agnostic but do require that the full data model be implemented by the codec.
+  * Reserved `_type` property.
+
+While the `_type` property is reserved in every map regardless of where it appears in a block, IPLD Generics **MUST** include the value of `_type` in the root property of a discrete block. Note: you can still embed multiple IPLD Generics in a single block using [inline blocks]().
+
+### Version 1
+
+The `_type` property is a string identifier. This identifier is used to lookup the implementation and if it cannot be found by the host environment any operation is expected to throw an exception. 
+
+Implementations MUST NOT fallback to *Layer 1* operations of the contents of the node.
+
+### Version 2
+
+The `_type` property is a `Map`.
+
+The map must contain the following properties.
+
+  * `name` must be a string identifier.
+  * `engine` must be one of the following:
+	* `”IPLD/Engine/WASM/0”`
+
+Each additional property describes the implementation of every operation.
+
+*TODO: define structure of pointers to WASM functions*
+
+## Operations
+
+This section describes only the operations for which there is currently an implementation. For a more exhaustive list of operations we may support in the future read [IPLD Multi-block Collections](https://github.com/ipld/specs/blob/master/schema-layer/data-structures/multiblock-collections.md).
+
+The term “leaf” is used often below. When an operation is performed on a node it may make any number of additional calls to operations on other data structures with many results being produced and passed into continuations. “leaf” responses are those produced by an engine that are from the original target operation and not from subsequent operations trigger by the original operation.
+
+### GET `{ path }`
+
+Used for property lookups. Enables functionality similar to what you would expect from `Map` and `List` kinds.
+
+Only a single leaf `{ result: Value }` must be received.
+
+### KEYS `{}`
+
+Use to produce an iterator of every top level property available in the data structure. Takes no options.
+
+Multiple leaf `{ result: List(...keys) }` may be received.
+
+### READ `{ start, end }`
+
+Reads binary data.
+
+If `start` is omitted, default to 0.
+
+If `end` is omitted, default to the end of binary representation of the data structure.
+
+Multiple leaf `{ result: Bytes }` may be received.
+
+## Implementations
+
+While each implementation has a language specific API there is a fair amount in common between implementations that is specified here.
+
+The interface for every operation can do the following:
+
+  * Return a result.
+  * Perform multiple calls of **any** operation on multiple targets of `value` or `path`
+	  *  `value` is an IPLD Data Model value.
+	  *  `path` is a IPLD Path relative to the node performing the operation.
+		  *  This path should be interpreted as *Layer 1* within the node and *Layer 2* once the path leaves the current block.
```

Comment:
bad indenting

Reply:
I think you might need to expand on what this means, it's not even clear to me. It seems to me that "Layer 1" and "Layer 2" are the wrong way around here since "within the node", assuming a Typed node, it's got to be interpreted in some special way. Although for a `GET` is there another way to interpret a path than pull off the first segment of a `'/'` separated string and do something non-layer-1 with it? Does that need to be made clear?

Reply:
Ya, we don’t have a good definition of “path” in terms of layers, but I also didn’t have a better way to describe it.

Basically, you need to read into the node value (decoded block data) until you hit a link, then you need to run the rest of the path through the generics engine. You can’t use the generic implementation for the read because it could be a `get` method itself that asked for the read.

Note: this is another reason generic root nodes must be discrete blocks.